Allow for emission of swift.extension symbols for extensions to external types in SymbolGraphGen

theMomax opened the pull request:
This PR introduces a new mode to the Symbol Graph generation that changes how extensions to external types are handled. Since we are explicitly only talking about extensions to **external** types, all symbols and relationships under consideration live in Symbol Graph Files with names of the form `A@B.symbols.json`. 

Any symbols resulting from such an extension are no longer directly associated with the extended type (which does not exist in the same Symbol Graph File), but with a new kind of symbol, the `swift.extension` symbol. There is one such symbol for every `extension` block that extends an external type. This `swift.extension` symbol is also connected to the symbol of the type it extends via an `extensionTo` relationship.

These changes are directly related to this [discussion about the Symbol Graph Format](https://forums.swift.org/t/symbol-graph-adaptions-for-documenting-extensions-to-external-types-in-docc/56684). This [initial pitch](https://forums.swift.org/t/document-extensions-to-external-types-using-docc/56060) motivates the feature and gives more context on why these changes are necessary.

List of changes:
 - introduction of the `swift.extension` symbol and `extensionTo` relationship
 - adding support for `ExtensionDecl` to the `Symbol` class
 - adding a `typeKind` field to the symbol's extension mixin which indicates what kind of
   symbol was extended
 - introduction of the `-emit-extension-block-symbols` flag, which enables the behavior
   outlined below
 - adaptions to `SymbolGraphASTWalker` that ensure a `swift.extension` symbol is emitted
   for each extension to a type that does not exist in the local symbol graph
 - adaptions to `SymbolGraph` and `SymbolGraphASTWalker` that ensure relationships are
   correctly associated with the `swift.extension` symbol instead of the original type
   declaration's (extended nominal's) symbol where applicable
 - adaptions to `SymbolGraphASTWalker` that ensure `swift.extension` symbols are connected
   to their respective extended nominal's symbol using an `extensionTo` relationship

Overall, these changes should ensure that:
 - documentation frameworks can access the documentation comments on top of extension blocks for extensions to external types
 - documentation frameworks can reconstruct which members/conformances are associated with a certain extension block (and its documentation)
 - documentation frameworks can know what kind of type was extended (struct/class/enum/protocol)
 - legacy documentation frameworks don't observe a breaking change (as all breaking changes are hidden behind the CLI flag)

This PR can be seen as a first step to fixing [SR-15410](https://github.com/apple/swift-docc/issues/210), but more changes to other repositories further up the stack will be necessary.

theMomax committed:
Allow for emission of swift.extension symbols for extensions to external types in swiftSymbolGraphGen

This includes:
 - bumping the SWIFT_SYMBOLGRAPH_FORMAT_MINOR version
 - introduction of the "swift.extension" symbol and "extensionTo" relationship
 - adding support for ExtensionDecl to the Symbol class
 - adding a "typeKind" field to the symbol's extension mixin which indicates what kind
   of symbol was extended
 - intoduction of the -emit-extension-block-symbols flag, which enables the behavior
   outlined below
 - adaptions to SymbolGraphASTWalker that ensure a swift.extension symbol is emitted
   for each extension to a type that does not exist in the local symbol graph
 - adaptions to SymbolGraph and SymbolGraphASTWalker that ensure member and conformance
   relationships are correctly associated with the swift.extension symbol instead of
   the original type declaration's (extended nominal's) symbol where applicable
 - adaptions to SymbolGraphASTWalker that ensure swift.extension symbols are connected
   to their respective extended nominal's symbol using an extensionTo relationship

Testing:
- adds SymbolGraph tests that test behavior only relevant in
  -emit-extension-block-symbols mode
- adapts some SymbolGraph tests to additionally test similar behavior for
  extensions to external types in -emit-extension-block-symbols mode
- adapts some SymbolGraph tests to (additionally or exclusively) test the
  behavior with -emit-extension-block-symbols mode enabled

Bugfixes:
- fixes a bug where some conformsTo relationships implicated by the conformances
  declared on an extension to an external type were not emitted
  (see test/SymbolGraph/Relationships/ConformsTo/Indirect.swift)

Further changes:
- documents the strategy for naming and associating children declared in extensions
  to typealiases (see test/SymbolGraph/Relationships/MemberOf/Typealias.swift,
  test/SymbolGraph/Symbols/Names.swift)

daniel-grumberg commented on include/swift/Option/Options.td:
```diff
@@ -1326,6 +1326,11 @@ def pretty_print: Flag<["-"], "pretty-print">,
   Flags<[SwiftAPIExtractOption, SwiftSymbolGraphExtractOption]>,
   HelpText<"Pretty-print the output JSON">;
 
+def emit_extension_block_symbols: Flag<["-"], "emit-extension-block-symbols">,
+  Flags<[SwiftSymbolGraphExtractOption, FrontendOption,
+        NoInteractiveOption, SupplementaryOutput, HelpHidden]>,
+  HelpText<"Emit 'swift.extension' symbols for extensions to external types instead of directly associating members and conformances with the extended nominal">;
+
```
Ideally we would want to make this behaviour the default so usability of the flag is not a major concern, so it might be worth indicating in the flag name itself that this is a symbol graph extraction only feature. Maybe `-emit-symbol-graph-with-extension-block-symbols`. At least the help text should definitely mention symbol graphs otherwise it isn't very clear that this is a symbol graph feature.

theMomax replied:
Good catch, thank you...I guess I got a little lost in my own world there!

daniel-grumberg replied:
I think it needs to be more explicit than this.

daniel-grumberg commented on lib/DriverTool/swift_symbolgraph_extract_main.cpp:
```diff
@@ -171,7 +171,7 @@ int swift_symbolgraph_extract_main(ArrayRef<const char *> Args,
       ParsedArgs.hasArg(OPT_include_spi_symbols),
       /*IncludeClangDocs=*/false,
-  };
+      ParsedArgs.hasArg(OPT_emit_extension_block_symbols)};
```
```suggestion
      ParsedArgs.hasArg(OPT_emit_extension_block_symbols)
    };
```

theMomax replied:
Interestingly, you need to use trailing comma syntax, otherwise clang-format will revert your suggested change to what  I committed.

daniel-grumberg commented on lib/SymbolGraphGen/JSON.cpp:
```diff
@@ -72,6 +72,27 @@ void swift::symbolgraphgen::serialize(const ExtensionDecl *Extension,
       }
+
+      StringRef Kind;
+      switch (ExtendedNominal->getKind()) {
```
Is there no existing API for doing this? as in converting a `DeclKind` to a string suitable for symbol graph output? If not this would be a good opportunity to refactor this code.

daniel-grumberg replied:
Please include the `swift.` prefix. I will at some point add similar functionality to clang for Objective-C categories, when we do that it might make sense to unify the SGF representation for both concepts since they are extremely similar.
WRT refactoring this with `Symbol.getKind(Decl *)`, I had a look at that function and it feels like it should be a `static` in Symbol as opposed to a member function. If you do that it might make sense to just make this chunk of code something like:
```
switch (ExtendedNominal->getKind()) {
    case DeclKind::Class:
    case DeclKind::Struct:
    case DeclKind::Enum:
    case DeclKind::Protocol:
        Kind = Symbol::getKind(ExtendedNominal);
        break;
    default:
        Kind = "unknown"
        break;
}
```

theMomax replied:
Resolved in 9af13ab80bdb660e89f70b6c9dd64171bcea378e. I didn't add the extra switch filter in JSON.cpp as I think the JSON serialization shouldn't have authority over what kinds are valid in the context of `swiftExtension.typeKind`.

daniel-grumberg commented on lib/SymbolGraphGen/Symbol.cpp:
```diff
@@ -597,52 +634,70 @@ void Symbol::serialize(llvm::json::OStream &OS) const {
-          Decl = Nominal;
-        } else {
-          Decl = dyn_cast_or_null<ValueDecl>(DC->getAsDecl());
+  auto collectPathComponents =
```
was this formatted with clang format? In general it is best to do formatting with git clang-format

theMomax replied:
Sorry about that. I always formatted as suggested in the [guide](https://github.com/apple/swift/blob/main/docs/HowToGuides/FAQ.md#how-do-i-format-my-changes), so I'm not sure what happened there exactly. However, if I reformat the entire document now, it keeps the formatting as I committed it before. Thus, I guess git clang-format noticed the faulty indentation in one of the lines I edited and then expanded the re-format to the whole function-definition. Only changing the indentation of one line wouldn't make any sense either.

daniel-grumberg replied:
It's completely fine, I forgot that it keeps a bit of context around the modifications.

theMomax replied:
Resolved in 37f4c983c382991edd2c776e98250a9ce3a96d0e. I had to add a little more type assertions to make your code compile, though.

daniel-grumberg commented on lib/SymbolGraphGen/Symbol.cpp:
```diff
@@ -710,8 +765,30 @@ bool Symbol::supportsKind(DeclKind Kind) {
   case DeclKind::Subscript: LLVM_FALLTHROUGH;
   case DeclKind::TypeAlias: LLVM_FALLTHROUGH;
   case DeclKind::AssociatedType:
+    LLVM_FALLTHROUGH;
```
fall through on previous line just for consistency

QuietMisdreavus reviewed (commented):
This looks really nice, thanks! I have a couple of minor comments. Was there anything you wanted to add to this implementation before building on it in SymbolKit and Swift-DocC?

Also, some of the test files in this PR are missing the end-of-file newline. GitHub labels this as a little ⛔ icon at the end of the file. This shouldn't affect the tests building and passing, but Git (and other tools that expect it, like Vim) will make a note of it when looking at them.

theMomax commented on lib/SymbolGraphGen/SymbolGraphASTWalker.cpp:
```diff
@@ -173,30 +190,35 @@ bool SymbolGraphASTWalker::walkToDeclPre(Decl *D, CharSourceRange Range) {
-        const auto *Comp = UnexpandedCompositions.pop_back_val();
-        for (const auto &Member : Comp->getMembers()) {
-          HandleProtocolOrComposition(Member);
+      while (!UnexpandedCompositions.empty() || !UnexpandedProtocols.empty()) {
```
Maybe I misjudged the situation, but I don't think this can be split up into two loops.

What I'm trying to achieve here is to record `conformsTo` relationships between the extension block symbol and any conformances added via this extension. However, we do not only want to include those protocols that are explicitly stated on the extension block, _but also implicitly inherited ones_. To do so, we have to expand two different constructions:

Protocol declarations (`UnexpandedProtocols`), which state one or more inheritances, e.g.:
```swift
public protocol X: A, B {
    // ...
}
```
Protocol compositions (`UnexpandedCompositions`), e.g.:
```swift
public typealias Y = C & D
```

I assume that these two constructions can be nested indefinitely, e.g. like this:

```swift
public protocol A: B, C { }

public typealias B = D & E

public protocol D: F, G { }

public typealias F = H & I

// ...
```

If I were to break up the loop into two separate ones, I couldn't handle this situation, could I?

theMomax replied:
Resolved in bb78034edf9d004b260cf890998699faab7dffba.

daniel-grumberg replied:
I still think this is unintuitive control flow, it took me a while to figure out how this was working. Would it not make more sense to keep track only of unexpanded protocols and handle all compositions eagerly? Essentially what I amsuggesting is to make this just an iteration over `UnexpandedProtocols` and turn `HandleProtocolOrComposition` into:
```
// Add the type to the right list for expansion.
      auto HandleProtocolOrComposition = [&](Type InitialTy) {
        SmallVector<const ProtocolCompositionType *, 4> UnexpandedLocalCompositions = { InitialTy };
        while (const auto *Ty = UnexpandedLocalCompositions.pop_back_val()) {
            if (const auto *Proto = dyn_cast_or_null<ProtocolDecl>(Ty->getAnyNominal())) {
                UnexpandedProtocols.push_back(Proto);
            } else if (const auto *Comp = Ty->getAs<ProtocolCompositionType>()) {
                 UnexpandedLocalCompositions.append(Comp->getMembers());
            } else {
               abort();
            }
      };
```
With this you can get rid of `UnexpandedCompositions` at the outer scope and make this loop pretend like it only ever deals with `UnexpandedProtocols` which seems more intuitive.

theMomax replied:
Resolved in 37f4c983c382991edd2c776e98250a9ce3a96d0e. It turned out a little more verbose, though, as I had to add some type assertions in the beginning to make the code compile.

theMomax commented on lib/SymbolGraphGen/SymbolGraph.cpp:
```diff
@@ -241,12 +238,31 @@ void SymbolGraph::recordMemberRelationship(Symbol S) {
+              dyn_cast_or_null<ExtensionDecl>(DC->getAsDecl())) {
+        auto const shouldBeRecordedAsExtension =
+            AssociateMembersWithExtensionBlock &&
+            !Extension->getModuleContext()->getNameStr().equals(
```
If I understood the documentation of your PR correctly, then the problem looked as follows:

Assuming we have a Swift module `A` with an underlying Clang module, then the Swift compiler handles the underlying Clang module by wrapping it in a Swift module also called `A`, which is then included in the Swift module `A` using an `@_exported import A` statement.

This then caused problems in SymbolGraphGen because the (C) `A` was known to be `@_exported` imported and the module comparison was exclusively done via the name, so everything in (Swift) `A` was also treated as if it were `@_exported` imported. Thus, extensions to external types such as `public extension String {}` were also exported to the MainGraph.

Is that about right?

Now, this is how I think that problem translates to my case:

The only case where the name equality comparison designates two different modules as equal is if it compares the (Swift) `A` with the (C) `A`. In my logic, this would be the case if (Swift) `A` extended a type in (C) `A`. This is possible. In that case, the string equality would consider the Swift/C modules equal and thus `shouldBeRecordedAsExtension` is set to `false`. This is consistent with the fact that the contents of this extension are stored in `A.symbols.json`, not `A@A.symbols.json` (which does not exist).

Is this the case you were concerned about? Does the behavior described above match what you expected to happen?

I tried to capture this behavior in [test/SymbolGraph/ClangImporter/ClangExtensions.swift](https://github.com/apple/swift/blob/89667d43306780922702ff5ce10331722177d179/test/SymbolGraph/ClangImporter/ClangExtensions.swift).

QuietMisdreavus commented:
@swift-ci Please smoke test

daniel-grumberg commented on lib/SymbolGraphGen/Symbol.cpp:
```diff
@@ -709,9 +767,30 @@ bool Symbol::supportsKind(DeclKind Kind) {
   }
 }
+
+AccessLevel Symbol::getEffectiveAccessLevel(const ExtensionDecl *ED) {
```
Shouldn't just `ED.getFormalAccess()` compute this? There is no way this computation doesn't exist anywhere else in the compiler. We shouldn't reinvent the wheel here.

theMomax replied:
No, unfortunately it doesn't. I think `ED.getMaxAccessLevel()` computes the maximum access level one of its members **can** have based on the extended type and types used in constraints.

I want the effective access level for this ExtensionDecl to be `internal`.
```swift
extension String {
    func foo() { }
}
```

I spent some time searching for a fitting function, but couldn't find one. It kinda makes sense as this logic isn't really relevant to the rest of the compiler, as the extension block itself isn't an accessible symbol in Swift code.

theMomax replied:
I extended the comment in Symbol.h accordingly in fc872b09428fc858852befb64bc4b10680a66d99.

daniel-grumberg commented on lib/SymbolGraphGen/SymbolGraph.cpp:
```diff
@@ -324,7 +340,13 @@ void SymbolGraph::recordConformanceSynthesizedMemberRelationships(Symbol S) {
-
+      
+      // If D is not the OwningNominal, it is an ExtensionDecl. In that case
+      // we only want to get members that were enabled my this exact extension.
```
nit s/my/by

daniel-grumberg commented on lib/SymbolGraphGen/SymbolGraphASTWalker.cpp:
```diff
@@ -159,25 +158,65 @@ bool SymbolGraphASTWalker::walkToDeclPre(Decl *D, CharSourceRange Range) {
+    // local types are directly associated with the extended nominal.
+    auto const shouldBeRecordedAsExtension =
+        Options.EmitExtensionBlockSymbols &&
+        !Extension->getModuleContext()->getNameStr().equals(
```
This part of the expression is also used in SymbolGraph.cpp, it might be worth to refactor this into it's own function (maybe be a free function in the `symbolgraphgen` namespace).

theMomax replied:
Resolved in c20c488e25f63a774c80292a95e0a67325dda154. I added the function to `SymbolGraphASTWalker` as that stores the `SymbolGraphOptions` which are required to evaluate the condition.

daniel-grumberg commented on lib/SymbolGraphGen/SymbolGraphASTWalker.cpp:
```diff
@@ -159,25 +158,65 @@ bool SymbolGraphASTWalker::walkToDeclPre(Decl *D, CharSourceRange Range) {
     // If there are some protocol conformances on this extension, we'll
     // grab them for some new conformsTo relationships.
     if (!Extension->getInherited().empty()) {
+      // We want to add conformsTo relationships for all protocols implicitly
```
Thanks for the comment! It's very useful.

daniel-grumberg reviewed (commented):
This looks pretty good at this point. Some smaller issues left to consider.

daniel-grumberg commented:
I am happy with the changes at this point. Can we rebase this, so that we can run CI?

theMomax commented:
> I am happy with the changes at this point. Can we rebase this, so that we can run CI?

I've been working off of tag `swift-DEVELOPMENT-SNAPSHOT-2022-08-30-a` recently because checking out main fails for me (it used to work, though):

```
> utils/update-checkout --scheme main
... omitted some lines here ...
[llvm-project]                          + git checkout stable/20220421
[llvm-project]                          Previous HEAD position was 57827dcce6fa Merge pull request #5224 from akyrtzi/pr/stable/ignore-module-flags-if-not-enabled
[llvm-project]                          Switched to branch 'stable/20220421'
[llvm-project]                          Your branch is up to date with 'origin/stable/20220421'.
[swift-nio-ssl]                         + git submodule update --recursive
[sourcekit-lsp]                         + git submodule update --recursive
[swift-experimental-string-processing]  + git submodule update --recursive
[llvm-project]                          + git fetch --recurse-submodules=yes --tags
[llvm-project]                          + git rebase FETCH_HEAD
[llvm-project]                          Current branch stable/20220421 is up to date.
[llvm-project]                          + git submodule update --recursive
======UPDATE FAILURES======
Traceback (most recent call last):
  File "/Users/themomax/Development/github.com/apple/swift/utils/update-checkout", line 15, in <module>
    update_checkout.main()
  File "/Users/themomax/Development/github.com/apple/swift/utils/update_checkout/update_checkout/update_checkout.py", line 639, in main
    fail_count += check_parallel_results(update_results, "UPDATE")
  File "/Users/themomax/Development/github.com/apple/swift/utils/update_checkout/update_checkout/update_checkout.py", line 70, in check_parallel_results
    print("%s failed (ret=%d): %s" % (r.repo_path, r.ret, r))
AttributeError: 'KeyError' object has no attribute 'repo_path'
```

theMomax commented:
@daniel-grumberg I managed to fix the issue on my side and rebased. You can run the CI now!

daniel-grumberg commented:
@swift-ci Please smoke test

QuietMisdreavus commented on test/SymbolGraph/Symbols/Unavailable.swift:
```diff
@@ -26,3 +27,15 @@ extension ShouldAppear {
 
 // CHECK-NOT: shouldntAppear
+
+// CHECK-NOT: OSXUnavailable
```
Is this check line redundant? It's already being done earlier in the file.

QuietMisdreavus reviewed (approved):
Thanks so much! This looks ready to go for me; all the comments i left are pretty minor.

...however, it does look like i introduced another merge conflict. 😭 It should be pretty straightforward to fix, and once that's done and tested, i think we can land this!

theMomax commented on lib/SymbolGraphGen/SymbolGraphASTWalker.cpp:
```diff
@@ -159,25 +156,83 @@ bool SymbolGraphASTWalker::walkToDeclPre(Decl *D, CharSourceRange Range) {
+        if (const auto *Comp = Ty->getAs<ProtocolCompositionType>()) {
           UnexpandedCompositions.push_back(Comp);
         } else {
           abort();
```
Resolved.

theMomax commented:
Rebased, fixed conflicts, addressed your remaining comments, and squashed again. Thanks for the final review @QuietMisdreavus ! Can you please kick off the CI again?

QuietMisdreavus commented:
@swift-ci Please smoke test

QuietMisdreavus commented:
@theMomax Thanks so much! As soon as CI is green, i'll merge it.

d-ronnqvist commented:
Apologies for the late comment. Can we also add a new element in [lib/Option/features.json][1]? That help build systems know when they have a version of the Swift compiler that supports this new flag so that they don't need to hard code what version numbers support what flags. 

It doesn't have to be done in this PR if you prefer to land this now and open a new PR that update features.json.

[1]: https://github.com/apple/swift/blob/main/lib/Option/features.json

theMomax commented:
@d-ronnqvist if you don’t mind, I‘d prefer to update the features.json with my next PR, which will add support for default implementation relationships. Before that, the extension block symbol feature isn’t really „complete“ anyway. I cut that part out of this PR as the change set had become quite large already (on this PR and the accompanying swift-docc PR).

QuietMisdreavus commented:
I'll go ahead and merge this, since it's all ready to go and doesn't affect anything unless you add the new flag.
